Python/pylifecycle.c

```diff
@@ -1290,6 +1290,13 @@ finalize_interp_clear(PyThreadState *tstate)
         _PyGC_CollectNoFail();
     }
 
+    /* Clear all loghooks */
+    /* Both _PySys_ClearAuditHooks function and users still need PyObject,
+       such as tuple. */
+    if (is_main_interp) {
+        _PySys_ClearAuditHooks(tstate);
```

Comment:
It would be safer to call it after PyInterpreterState_Delete(). It requires to simplify _PySys_ClearAuditHooks() to only access _PyRuntime: don't call _PySys_Audit() anymore, and avoid `PyThreadState *` since it doesn't exist anymore after PyInterpreterState_Delete().

After PyInterpreterState_Delete() is called, it's no longer possible to execute Python code.

I suggest to move the call in Py_Finalize() after finalize_interp_delete() call.

Reply:
`_PySys_Audit` uses tuple object, which needs GC support.
If we move the call after `finalize_interp_delete`, we must redesign the `_PySys_Audit` which will be called in `_PySys_ClearAuditHooks`.

I think there is no chance for pure python code to execute code in this patch.

...Or maybe they change object in `finalize_interp_types` with ctypes?

Reply:
The hook can (and will) do anything, but the more important issue is that PySys_Audit will create a tuple object, which can't be allowed after finalization. And we can't change the API to allow passing `NULL` for `args` at this stage.

After doing my own experimentation, this is the best place to put it.